File: regression/python/clique.py
from z3 import *

# We represent graphs as a list of pair of non-negative integers.

def clique(edges, n):
    assert all(map(lambda e: isinstance(e[0], int) and isinstance(e[1], int), edges))
    assert all(map(lambda e: e[0] >= 0 and e[1] >= 0, edges))
    num = 0
    for e in edges:
        if e[0] >= num:
            num = e[0] + 1
        if e[1] >= num:
            num = e[1] + 1
    assert num <= 128, "too many nodes for demo"
    S = Solver()
    # Create a 0-1 variable in_i for each node, in_i is 1 if node i is in the clique
    In = [ Int('in_%s' % i) for i in range(num) ]
    for i in range(num):
        S.add(0 <= In[i], In[i] <= 1)
    # The number of nodes in the clique must be greater or equal to n
    S.add(simplify(reduce(lambda a, b: a + b, In, 0)) >= n)
    # Create a matrix from the edges
    matrix = [ [ 0 for j in range(i)] for i in range(num)]
    for e in edges:
        s = e[0]
        t = e[1]
        if s < t:
            s = e[1]
            t = e[0]
        matrix[s][t] = 1
    # For each non-existing edge (i, j) Add constraint In[i] == 0 or In[j] == 0
    for i in range(num):
        for j in range(i):
            if i != j:
                if matrix[i][j] == 0:
                    S.add(Or(In[i] == 0, In[j] == 0))
    r = S.check()
    if r == unsat:
        print("no solution")
    elif r == unknown:
        print("failed")
        print S.reason_unknown()
    else:
        print("found solution")
        m = S.model()
        print filter(lambda a: eq(m.eval(In[a]), IntVal(1)), range(num))

clique([(0, 1), (2, 3), (1, 2), (2, 3), (1, 4), (1, 5), (2, 5)], 3)
clique([(0, 1), (2, 3), (1, 2), (2, 3), (1, 4), (1, 5), (2, 5)], 4)


            

File: regression/cs/clique.cs
using Microsoft.Z3;
using System;
using System.Collections.Generic;

class Test
{
    public class Edge
    {
        public uint v0, v1;
        public Edge(uint x, uint y) { v0 = x; v1 = y; }
    }

    protected void Clique(Context ctx, Edge[] edges, uint n)
    {
        uint num = 0;
        foreach (Edge e in edges)
        {
            if (e.v0 >= num)
                num = e.v0 + 1;
            if (e.v1 >= num)
                num = e.v1 + 1;
        }

        Solver S = ctx.MkSolver();

        IntExpr [] In = new IntExpr[num];

        for (uint i = 0; i < num; i++)
        {
            In[i] = ctx.MkIntConst(String.Format("in_{0}", i));
            S.Assert(ctx.MkLe(ctx.MkInt(0), In[i]));
            S.Assert(ctx.MkLe(In[i], ctx.MkInt(1)));
        }

        ArithExpr sum = ctx.MkInt(0);
        foreach (IntExpr e in In)
            sum = ctx.MkAdd(sum, e);
        S.Assert(ctx.MkGe(sum, ctx.MkInt(n)));
        
        IntNum[][] matrix = new IntNum[num][];

        for (uint i = 0; i < num; i++)
        {
            matrix[i] = new IntNum[i];
            for (uint j = 0; j < i; j++)
                matrix[i][j] = ctx.MkInt(0);
        }

        foreach (Edge e in edges)
        {
            uint s = e.v0;
            uint t = e.v1;
            if (s < t) {
                s = e.v1;
                t = e.v0;
            }
            matrix[s][t] = ctx.MkInt(1);
        }

        for (uint i = 0; i < num; i++)
            for (uint j = 0; j < i; j++)
                if (i != j)
                    if (matrix[i][j].Int == 0)
                        S.Assert(ctx.MkOr(ctx.MkEq(In[i], ctx.MkInt(0)), 
                                          ctx.MkEq(In[j], ctx.MkInt(0))));

        Status r = S.Check();
        if (r == Status.UNSATISFIABLE)
            Console.WriteLine("no solution");
        else if (r == Status.UNKNOWN)
        {
            Console.Write("failed");
            Console.WriteLine(S.ReasonUnknown);
        }
        else
        {
            Console.WriteLine("solution found");
            Model m = S.Model;
            
            Console.Write("{ ");
            foreach (FuncDecl cfd in m.ConstDecls)
            {
                IntNum q = (IntNum)m.ConstInterp(cfd);
                if (q.Int == 1)
                    Console.Write(" " + cfd.Name);
            }
            Console.WriteLine(" }");

            Console.Write("{ ");
            for (uint i = 0; i < num; i++)
            {
                IntNum q = (IntNum)m.Evaluate(In[i]);                
                if (q.Int == 1)
                    Console.Write(" " + In[i]);
            }
            Console.WriteLine(" }");
        }

    }

    public void Run()
    {
        Dictionary<string, string> cfg = new Dictionary<string, string>() { 
            { "AUTO_CONFIG", "true" } };

        using (Context ctx = new Context(cfg))
        {
            Edge[] g = { new Edge(0, 1), new Edge(2, 3), new Edge(1, 2), new Edge(2, 3), new Edge(1, 4), new Edge(1, 5), new Edge(2, 5) };

            Clique(ctx, g, 3);
            Clique(ctx, g, 4);
        }
    }
}